### PR TITLE
Generate valid container id in fake docker client.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/kubernetes/kubernetes/issues/41488.

This PR generates valid container id by hashing container name. So that dockershim checkpoint won't report `checkpoint key XXX is not valid` error.

This PR also reverts #41460.

@yujuhong @freehan /cc @kubernetes/sig-node-pr-reviews

### CODE REVIEW

**Comment by Random-Liu** [comment]
@k8s-bot gci gke e2e test this

**Comment by yujuhong** [comment]
@wojtek-t @gmarek is there any kubemark test we can trigger for the PR to actually test this? The existing ones don't start any pods.

**Comment by Random-Liu** [comment]
@yujuhong It seems not. I'll run it myself. :)

**Comment by feiskyer** [comment]
LGTM.

@k8s-bot cross build this

**Comment by wojtek-t** [comment]
@yujuhong @Random-Liu - with https://github.com/kubernetes/kubernetes/pull/41490 (that has been merged few hours ago), the kubemark presubmit script will actually start a single pod (well a pod from an RC) and will wait until it will be running and ready. So right now kubemark presubmit should be enough.

**Comment by Random-Liu** [comment]
@k8s-bot kubemark test this.

@wojtek-t Could you approve the PR? Thanks!

**Comment by wojtek-t** [comment]
@Random-Liu - it seems that something still doesn't work.

**Comment by Random-Liu** [comment]
@yujuhong Updated the PR. This time it should work. Let's wait for the test result.

**Comment by yujuhong** [comment]
/lgtm

Thanks!

**Comment by Random-Liu** [comment]
@yujuhong There still seems to be something wrong...Let's wait for the test result. :P

**Comment by Random-Liu** [comment]
@k8s-bot test this.

**Comment by yujuhong** [comment]
@wojtek-t @gmarek @shyamjvs we can't mock out the OS interface in dockershim anymore (given that it should not be part of kubelet). This means that hollow nodes will still create directories, checkpoints, and symlinks. The fake docker client has also become more and more complex since it needs to fake more information (log path, labels, etc). More reasons to switch to using a fake CRI client :)

**Comment by Random-Liu** [comment]
@yujuhong The PR is finally ready. I updated the fake docker client to be able to list containers with filters, and also updated the dockershim not to create symlink when the returned log path is empty.

All those are quite crispy. We should move to fake CRI client.

File: pkg/kubelet/dockertools/docker_manager_linux_test.go [hunk 1]
```diff
@@ -260,13 +236,7 @@ func TestSecurityOptsAreNilWithDockerV19(t *testing.T) {
 		"create", "start", "inspect_container",
 	})
 
-	fakeDocker.Lock()
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/docker_manager_linux_test.go]: Nice cleanup!

**Review by yujuhong — COMMENTED** [review]
/lgtm

**Comment by Random-Liu** [comment]
Apply LGTM based on https://github.com/kubernetes/kubernetes/pull/41524#pullrequestreview-22613824

**Comment by liggitt** [comment]
Pretty sure this broke kubemark: https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/

**Comment by Random-Liu** [comment]
> Pretty sure this broke kubemark: https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/

Yeah, this only broke kubemark-500. It's really weird, there is no any debug information. :(

**Pull request merged**